### PR TITLE
Add BLOCK_LOAD_STRIPED and BLOCK_STORE_STRIPED

### PR DESCRIPTION
**Pull request by mnicely**
This PR adds the following to `BlockLoadAlgorithm`
1. BLOCK_LOAD_STRIPED
It's basically BLOCK_LOAD_TRANSPOSE without the BlockExchange

This PR adds the following to `BlockStoreAlgorithm`
1. BLOCK_STORE_STRIPED
It's basically BLOCK_STORE_TRANSPOSE without the BlockExchange

### CODE REVIEW

**Comment by mnicely** [comment]
@lllukas & @jszuppe for vis

**Comment by allisonvacanti** [comment]
This looks good, I'll plan to take a closer look sometime next week.

Can you update the `test_block_load_store.cuh` test? It should just take an extra line in [this block](https://github.com/NVIDIA/cub/blob/main/test/test_block_load_store.cu#L439-L441) that exercises the new `BlockAlgorithm`.

**Comment by mnicely** [comment]
Thanks @allisonvacanti. Looks like everything passes.

```bash
...
      Start  3: cub.cpp14.test.block_load_store.thorough
 3/52 Test  #3: cub.cpp14.test.block_load_store.thorough ..............   Passed    0.14 sec
      Start  4: cub.cpp14.test.block_load_store.benchmark
 4/52 Test  #4: cub.cpp14.test.block_load_store.benchmark .............   Passed    0.10 sec
...
```

This PR covers our C2C scenarios. I need to think more about how R2C and C2R might work

**Comment by allisonvacanti** [comment]
DVS CL: 29946764
gpuCI: NVIDIA/thrust#1427

**Comment by allisonvacanti** [comment]
gpuCI identified some issues building with nvcc+clang. @mnicely can you take a look? Looks like a missing member variable.

```
../dependencies/cub/cub/block/block_store.cuh(628): error #245: a nonstatic member reference must be relative to a specific object
          detected during:
            instantiation of "void cub::BlockStore<T, BLOCK_DIM_X, ITEMS_PER_THREAD, ALGORITHM, BLOCK_DIM_Y, BLOCK_DIM_Z, PTX_ARCH>::StoreInternal<cub::BLOCK_STORE_STRIPED, DUMMY>::Store(OutputIteratorT, T (&)[ITEMS_PER_THREAD], int) [with T=char, BLOCK_DIM_X=15, ITEMS_PER_THREAD=1, ALGORITHM=cub::BLOCK_STORE_STRIPED, BLOCK_DIM_Y=1, BLOCK_DIM_Z=1, PTX_ARCH=800, DUMMY=0, OutputIteratorT=cub::DiscardOutputIterator<std::ptrdiff_t>]" 
(1048): here
            instantiation of "void cub::BlockStore<T, BLOCK_DIM_X, ITEMS_PER_THREAD, ALGORITHM, BLOCK_DIM_Y, BLOCK_DIM_Z, PTX_ARCH>::Store(OutputIteratorT, T (&)[ITEMS_PER_THREAD], int) [with T=char, BLOCK_DIM_X=15, ITEMS_PER_THREAD=1, ALGORITHM=cub::BLOCK_STORE_STRIPED, BLOCK_DIM_Y=1, BLOCK_DIM_Z=1, PTX_ARCH=800, OutputIteratorT=cub::DiscardOutputIterator<std::ptrdiff_t>]" 
../dependencies/cub/test/test_block_load_store.cu(137): here
            instantiation of "void Kernel<BLOCK_THREADS,ITEMS_PER_THREAD,LOAD_ALGORITHM,STORE_ALGORITHM,InputIteratorT,OutputIteratorT>(InputIteratorT, OutputIteratorT, OutputIteratorT, int) [with BLOCK_THREADS=15, ITEMS_PER_THREAD=1, LOAD_ALGORITHM=cub::BLOCK_LOAD_STRIPED, STORE_ALGORITHM=cub::BLOCK_STORE_STRIPED, InputIteratorT=const char *, OutputIteratorT=cub::DiscardOutputIterator<std::ptrdiff_t>]" 
../dependencies/cub/test/test_block_load_store.cu(180): here
            instantiation of "void TestKernel<T,BLOCK_THREADS,ITEMS_PER_THREAD,LOAD_ALGORITHM,STORE_ALGORITHM,InputIteratorT,OutputIteratorT>(T *, InputIteratorT, OutputIteratorT, OutputIteratorT, T *, T *, int, int) [with T=char, BLOCK_THREADS=15, ITEMS_PER_THREAD=1, LOAD_ALGORITHM=cub::BLOCK_LOAD_STRIPED, STORE_ALGORITHM=cub::BLOCK_STORE_STRIPED, InputIteratorT=const char *, OutputIteratorT=char *]" 
../dependencies/cub/test/test_block_load_store.cu(261): here
            instantiation of "void TestNative<T,BLOCK_THREADS,ITEMS_PER_THREAD,LOAD_ALGORITHM,STORE_ALGORITHM>(int, float, cub::Int2Type<1>) [with T=char, BLOCK_THREADS=15, ITEMS_PER_THREAD=1, LOAD_ALGORITHM=cub::BLOCK_LOAD_STRIPED, STORE_ALGORITHM=cub::BLOCK_STORE_STRIPED]" 
../dependencies/cub/test/test_block_load_store.cu(403): here
            instantiation of "void TestPointerType<T,BLOCK_THREADS,ITEMS_PER_THREAD,LOAD_ALGORITHM,STORE_ALGORITHM>(int, float) [with T=char, BLOCK_THREADS=15, ITEMS_PER_THREAD=1, LOAD_ALGORITHM=cub::BLOCK_LOAD_STRIPED, STORE_ALGORITHM=cub::BLOCK_STORE_STRIPED]" 
../dependencies/cub/test/test_block_load_store.cu(442): here
            instantiation of "void TestStrategy<T,BLOCK_THREADS,ITEMS_PER_THREAD>(int, float, cub::Int2Type<0>) [with T=char, BLOCK_THREADS=15, ITEMS_PER_THREAD=1]" 
../dependencies/cub/test/test_block_load_store.cu(476): here
            instantiation of "void TestItemsPerThread<T,BLOCK_THREADS>(int, float) [with T=char, BLOCK_THREADS=15]" 
../dependencies/cub/test/test_block_load_store.cu(491): here
            instantiation of "void TestThreads<T>(int, float) [with T=char]" 
../dependencies/cub/test/test_block_load_store.cu(534): here
```

File: cub/block/block_store.cuh [hunk 1]
```diff
@@ -576,6 +588,49 @@ private:
...
+
+        /// Store items into a linear segment of memory, guarded by range
+        template <typename OutputIteratorT>
+        __device__ __forceinline__ void Store(
+            OutputIteratorT   block_itr,                  ///< [in] The thread block's base output iterator for storing to
+            T                   (&items)[ITEMS_PER_THREAD], ///< [in] Data to store
+            int                 valid_items)                ///< [in] Number of valid items to write
+        {
+            if (linear_tid == 0)
+                temp_storage.valid_items = valid_items;     // Move through volatile smem as a workaround to prevent RF spilling on subsequent loads
```
Comment (allisonvacanti) [inline cub/block/block_store.cuh:628]: `temp_storage` needs to be cached as a member variable.

**Review by allisonvacanti — CHANGES_REQUESTED** [review]

**Comment by mnicely** [comment]
@allisonvacanti, I'll take a look!

**Comment by allisonvacanti** [comment]
@mnicely Any update on this? I'll be releasing 1.13 in a couple weeks, we should get this ready to test again sometime this week to make sure it can land in time.

**Comment by mnicely** [comment]
> @mnicely Any update on this? I'll be releasing 1.13 in a couple weeks, we should get this ready to test again sometime this week to make sure it can land in time.

Hi @allisonvacanti sorry for the delay!
I believe I have fixed the problem.

**Comment by canonizer** [comment]
This looks interesting. Is something like `BLOCK_LOAD_WARP_STRIPED` also planned? It would be the same as `BLOCK_LOAD_STRIPED`, but would load the data into the warp-striped arrangement instead.

**Comment by mnicely** [comment]
@canonizer I didn't need those for my use case, but I might not be hard to add...
@allisonvacanti If you have not objections, I'll see if I can get everything in today.

**Comment by mnicely** [comment]
I'm getting a bunch of errors trying to build CUB and test this go-around. Maybe it's best we wait till after 1.13 has been released and revisit.

```bash
...
/home/belt/workStuff/git_examples/cub/thrust/dependencies/cub/cub/iterator/tex_ref_input_iterator.cuh(133): error #1215-D: function "tex1Dfetch(texture<T, 1, cudaReadModeElementType>, int) [with T=uint2]"
/usr/local/cuda/bin/../targets/x86_64-linux/include/texture_fetch_functions.h(132): here was declared deprecated
          detected during:
            instantiation of "T cub::<unnamed>::IteratorTexRef<T>::TexId<UNIQUE_ID>::Fetch(Distance) [with T=TestBar, UNIQUE_ID=554, Distance=ptrdiff_t]" 
(299): here
            instantiation of "cub::TexRefInputIterator<T, UNIQUE_ID, OffsetT>::reference cub::TexRefInputIterator<T, UNIQUE_ID, OffsetT>::operator*() const [with T=TestBar, UNIQUE_ID=554, OffsetT=ptrdiff_t]" 
/home/belt/workStuff/git_examples/cub/thrust/dependencies/cub/test/test_iterator.cu(113): here
            instantiation of "void Kernel(InputIteratorT, T *, InputIteratorT *) [with InputIteratorT=cub::TexRefInputIterator<TestBar, 554, ptrdiff_t>, T=TestBar]" 
/home/belt/workStuff/git_examples/cub/thrust/dependencies/cub/test/test_iterator.cu(160): here
            instantiation of "void Test(InputIteratorT, T (&)[TEST_VALUES]) [with InputIteratorT=cub::TexRefInputIterator<TestBar, 554, ptrdiff_t>, T=TestBar, TEST_VALUES=8]" 
/home/belt/workStuff/git_examples/cub/thrust/dependencies/cub/test/test_iterator.cu(561): here
            instantiation of "void TestTexRef<T,CastT>() [with T=TestBar, CastT=TestBar]" 
/home/belt/workStuff/git_examples/cub/thrust/dependencies/cub/test/test_iterator.cu(694): here
            instantiation of "void Test<T,CastT>(cub::Int2Type<0>) [with T=TestBar, CastT=TestBar]" 
/home/belt/workStuff/git_examples/cub/thrust/dependencies/cub/test/test_iterator.cu(726): here
            instantiation of "void Test<T>() [with T=TestBar]" 
/home/belt/workStuff/git_examples/cub/thrust/dependencies/cub/test/test_iterator.cu(797): here
...
60 errors detected in the compilation of "/home/belt/workStuff/git_examples/cub/thrust/dependencies/cub/test/test_iterator.cu".
make[2]: *** [dependencies/cub/test/CMakeFiles/cub.cpp14.test.iterator.dir/build.make:76: dependencies/cub/test/CMakeFiles/cub.cpp14.test.iterator.dir/test_iterator.cu.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:18203: dependencies/cub/test/CMakeFiles/cub.cpp14.test.iterator.dir/all] Error 2
```

**Comment by allisonvacanti** [comment]
Hmm, strange -- that error should be suppressed by our build system. It's just a deprecation warning and can be ignored.

What platform are you seeing that on?

**Comment by mnicely** [comment]
I'm using CTK 11.3U1 and Ubuntu 20.04.

I've merged in the latest from main into my branch and I'm using the updated build instructions.

**Comment by mnicely** [comment]
Should one of these be ON?

```bash
THRUST_IGNORE_CUB_VERSION_CHECK:BOOL=OFF
THRUST_IGNORE_DEPRECATED_COMPILER:BOOL=OFF
THRUST_IGNORE_DEPRECATED_CPP_11:BOOL=OFF
THRUST_IGNORE_DEPRECATED_CPP_DIALECT:BOOL=OFF
```

**Comment by allisonvacanti** [comment]
No, those affect other things. I'll look into it, but that's a known issue that shouldn't hold things up, and 1.13 probably won't fix it.

Let me know if you are planning to add the other load algorithms Andy suggested, or if you want to leave that as future work. If you're finished with this patch for now, I'll restart the tests -- our CI seems to suppress the warning as intended, so we can just validate there.

**Comment by mnicely** [comment]
I think it'll be easier to hold off on `BLOCK_*_WARP_STRIDED`. 

We can revisit it later. I'll create a new PR and branch off 1.13 when it's released.

**Comment by allisonvacanti** [comment]
Sounds good 👍🏻  I'll go ahead and start the tests.

File: cub/block/block_store.cuh (hunk 1 above)
Comment (allisonvacanti) [reply cub/block/block_store.cuh:632]: That last commit only added an initialization to the constructor, there still needs to be a `temp_storage` member variable defined.

**Review by allisonvacanti — CHANGES_REQUESTED** [review]
Needs a bit more work -- @mnicely can you add the missing member variable?

```
TempStorage temp_storage;
```

needs to be added as a member to the new `StoreInternal` specialization.

File: cub/block/block_store.cuh (hunk 1 above)
Comment (mnicely) [reply cub/block/block_store.cuh:638]: Sorry, let me know if I got it correct this time.

**Comment by allisonvacanti** [comment]
Thanks! This looks good to me, I should be able to land it in time for 1.13 👍

**Comment by allisonvacanti** [comment]
DVS CL: 30058071
gpuCI: NVIDIA/thrust#1274

**Review by allisonvacanti — APPROVED** [review]

**Comment by allisonvacanti** [comment]
Tests look good, merging.

**Pull request merged**